### PR TITLE
OTT-1688: Parser Implementation for AdRule

### PR DESCRIPTION
**Pull request by Pubmatic-Dhruv-Sonone**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go [hunk 1]
```diff
@@ -4694,3 +4694,41 @@ func (o *OpenRTB) ORTBRegsGppSid() error {
 	}
 	return nil
 }
+
+// ORTBAdrule will read and set ortb Adrule parameter
+func (o *OpenRTB) ORTBAdrule() (err error) {
+
```
Comment (pm-nikhil-vaidya) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4700]: Can you please remove space

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go [hunk 2]
```diff
@@ -4694,3 +4694,41 @@ func (o *OpenRTB) ORTBRegsGppSid() error {
...
+	if !ok || err != nil {
+		return
+	}
+
+	reqExt := map[string]interface{}{}
+	if o.ortb.Ext != nil {
+		err = json.Unmarshal(o.ortb.Ext, &reqExt)
+		if err != nil {
+			return
+		}
```
Comment (pm-nikhil-vaidya) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4711]: Nitpick: We can combine this in single if condition

```go
if err := json.Unmarshal(o.ortb.Ext, &reqExt) ; err != nil {
   return
}
```

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go (hunk 1 above)
Comment (pm-nikhil-vaidya) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4700]: Nitpick: Can you please remove space here

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go [hunk 3]
```diff
@@ -4694,3 +4694,41 @@ func (o *OpenRTB) ORTBRegsGppSid() error {
...
+	video[ORTBExtAdrule] = val
+	wrapperExt[ORTBVideo] = video
+	reqExt[ORTBExtWrapper] = wrapperExt
+	data, err := json.Marshal(reqExt)
+	if err != nil {
+		return
+	}
+
+	o.ortb.Ext = data
+	return nil
```
Comment (pm-nikhil-vaidya) [inline modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4733]: Nitpick: we can directly return, no need to specify nil

**Review by pm-nikhil-vaidya — APPROVED** [review]
Approving, with some minor comments

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go (hunk 2 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4710]: ` o.ortb.Ext != nil ` is needed here to set `adrule`  even if  request.ext is not set.

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go (hunk 1 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go]: Done

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go (hunk 3 above)
Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go]: Done

File: modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go (hunk 2 above)
Comment (pm-nikhil-vaidya) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go:4710]: Only combine lines from 4708 to 4711, Not to remove the `o.ortb.Ext != nil` condition

Comment (Pubmatic-Dhruv-Sonone) [reply modules/pubmatic/openwrap/endpoints/legacy/ctv/parser_implementation.go]: Done

**Review by pm-nikhil-vaidya — DISMISSED** [review]

**Pull request closed without merging**